pkg/azuredisk/controllerserver.go

```diff
@@ -517,60 +525,165 @@ func (d *Driver) ListVolumes(ctx context.Context, req *csi.ListVolumesRequest) (
+	}
+
+	// get all resource groups and put them into a sorted slice
+	rgMap := make(map[string]bool)
+	volSet := make(map[string]bool)
+	for _, pv := range pvList.Items {
+		if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == DriverName {
+			diskURI := pv.Spec.CSI.VolumeHandle
+			if err := isValidDiskURI(diskURI); err != nil {
+				klog.Warningf("invalid disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			rg, err := GetResourceGroupFromURI(diskURI)
+			if err != nil {
+				klog.Warningf("failed to get resource group from disk uri (%s) with error(%v)", diskURI, err)
+				continue
+			}
+			volSet[diskURI] = true
+			if _, visited := rgMap[rg]; visited {
+				continue
+			}
+			rgMap[rg] = true
+		}
 	}
 
-	if start != 0 && start >= len(disks) {
+	if start != 0 && start > len(volSet) {
 		return nil, status.Errorf(codes.Aborted, "ListVolumes starting token(%d) on rg(%s) is greater than total number of volumes", start, d.cloud.ResourceGroup)
 	}
+	resourceGroups := make([]string, len(rgMap))
+	i := 0
+	for rg := range rgMap {
+		resourceGroups[i] = rg
+		i++
+	}
+	sort.Strings(resourceGroups)
 
-	maxEntries := len(disks) - start
-	if req.MaxEntries > 0 && int(req.MaxEntries) < maxEntries {
-		maxEntries = int(req.MaxEntries)
+	// loop through each resourceGroup to get disk lists
+	entries := []*csi.ListVolumesResponse_Entry{}
+	numVisited := 0
+	isCompleteRun := true
+	for _, resourceGroup := range resourceGroups {
+		if !isCompleteRun || (maxEntries > 0 && len(entries) >= maxEntries) {
+			isCompleteRun = false
+			break
+		}
+		listStatus := d.listVolumesByResourceGroup(ctx, resourceGroup, entries, start-numVisited, maxEntries-len(entries), &volSet)
+		numVisited += listStatus.numVisited
+		if listStatus.err != nil {
+			if status.Code(listStatus.err) == codes.FailedPrecondition {
+				continue
+			}
+			return nil, listStatus.err
+		}
+		entries = listStatus.entries
+		isCompleteRun = isCompleteRun && listStatus.isCompleteRun
+	}
+	nextTokenString := ""
+	if !isCompleteRun {
+		nextTokenString = strconv.Itoa(numVisited)
 	}
-	pageEnd := start + maxEntries
 
-	vmSet := d.cloud.VMSet
+	listVolumesResp := &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: nextTokenString,
+	}
+
+	return listVolumesResp, nil
+}
+
+// listVolumesInNodeResourceGroup is a helper function for ListVolumes used for when there is no available kubeclient
+func (d *Driver) listVolumesInNodeResourceGroup(ctx context.Context, start, maxEntries int) (*csi.ListVolumesResponse, error) {
 	entries := []*csi.ListVolumesResponse_Entry{}
-	for i := start; i < pageEnd; i++ {
-		d := disks[i]
-		nodeList := []string{}
+	listStatus := d.listVolumesByResourceGroup(ctx, d.cloud.ResourceGroup, entries, start, maxEntries, nil)
+	if listStatus.err != nil {
+		return nil, listStatus.err
+	}
+
+	nextTokenString := ""
+	if !listStatus.isCompleteRun {
+		nextTokenString = strconv.Itoa(listStatus.numVisited)
```

Comment:
I'm not sure if numVisited can work as the nextToken string in all the scenarios. 
Consider the scenario that I mentioned above when start = 6 and there's only 1 rg with 10 disks. The value of numVisited would be 4 and that will also be the next token string. When the call is made to ListVolumes the next time, it will be with startingToken=4  While actually it should be 10.